### PR TITLE
Otp 24 crypto fixes

### PR DESCRIPTION
**Pull request by kgautreaux**
Sorry for the delay pushing the branch.

As you can see the actual fix for the removed `:crypto.block_encrypt` and `:crypto.block_decrypt` in OTP 24 is literally _two lines_ in `aes.ex`, but the fixes caused a cascade of errors in the test suite.

I finally got some time the last few days to spelunk into the Erlang `:crypto` module and the NIF's in `aead.c`.  I relate the story here because I think it is interesting.

The test suite instructions [here](https://github.com/dwyl/fields#testing) currently recommend running the test suite with manually set env vars (`ENCRYPTION_KEYS="key1,key2" SECRET_KEY_BASE="key" mix test`), which must have worked fine in the `:crypto.block_encrypt` days, but after a lot of trial, error, and pretending I know how to read C I realized that in the OTP 24 era the cipher selection depends both on the atom passed to `:crypto.crypto_one_time_aead` **and** the _key size_.

So if you pass `:aes_256_gcm` as the cipher but you pass the `binary` "key1" as the key then your key size is only 24 bytes which corresponds to `:aes_192_gcm` not the `:aes_256_gcm` that you passed as your selected cipher and Erlang gives up and throws the`:badarg` exception saying that you have an `unknown cipher`.

So to fix the test suite failures we just have to make sure we pass a 32 byte key, but it seemed more ergonomic to me to try to inject the key as a dependency at run-time during the tests rather than rely on the tester passing the key with the right length as an env var.

That is why I changed the module attribute key extraction to a function so that it could be evaluated at run-time instead of compile time, and in the future an arbitrary key could be passed in during testing.

I think if I knew more about `ExUnit` this would have been a great idea, but instead I blundered my way into making a `ExUnit.CaseTemplate` using module as the basis for the testing so that common setup for the generated keys could be injected into each test module.  I'm still not sure this is a good solution, but here we are.

After all of that I figured I better actually commit something so I `@tag :skip`-ed the failing `hash` tests, which have hardcoded binary's for the comparison tests and don't tolerate the custom setup I introduced.

I apologize for the length I didn't have time to make it shorter.

### CODE REVIEW

**Comment by kgautreaux** [comment]
I had this CI problem locally as well. `Fields.TestCase` module wouldn't be available to `use` in the test files because it never got compiled to `.beam` code.  I thought it might be because it wasn't postfixed with `_test.exs` but changing that didn't cause it to get picked up by the compiler.  Eventually I had to `elixirc` the `.ex` file and move it into the test `/ebin` dir which I'm sure is a terrible hack because I don't understand the BEAM  build structure like I should. I would love to learn why elixir scripts in the test dir aren't being added to the module hierarchy during a test run.

**Comment by nelsonic** [comment]
Code changes LGTM. 👍 
Having a bit of a roadblock installing `Elixir @ 1.12.1` with OTP `24` on Mac. 🙄 
Will take a look with a fresh pair of eyes.

**Comment by th0mas** [comment]
@kgautreaux using `ExUnit.CaseTemplate` makes sense here, it's what it's designed for :) Your right in that it needs to be an `.ex` file - you need to add it to your compile path by changing `mix.exs`:

```diff
@@ -10,6 +10,7 @@ defmodule Fields.MixProject do
       start_permanent: Mix.env() == :prod,
       deps: deps(),
       package: package(),
+      elixirc_paths: elixirc_paths(Mix.env()),
 
@@ -28,6 +29,9 @@ defmodule Fields.MixProject do

+  defp elixirc_paths(:test), do: ["lib", "test/support"]
+  defp elixirc_paths(_), do: ["lib"]
```

I'll have a look at why Travis having issues as well...

**Comment by th0mas** [comment]
Added the environment variables to a config file and re-used the previous testing values.

@nelsonic Should be good to merge now 👍

File: lib/aes.ex [hunk 1]
```diff
@@ -86,6 +79,19 @@ defmodule Fields.AES do
   # """ # doc commented out because https://stackoverflow.com/q/45171024/1148249
   @spec get_key(number) :: String
   defp get_key(key_id) do
-    Enum.at(@keys, key_id)
+    Enum.at(fetch_keys(), key_id)
+  end
+
+  # Dependency injection of encryption keys to allow flexibility in tests and consumers.
+  # I don't think fetching the keys from the env is going to be too slow, but
+  # consider optimizing if benchmarking shows it is.
```
Comment (nelsonic) [inline lib/aes.ex:87]: reading: http://stratus3d.com/blog/2020/09/30/performance-of-elixirs-system-get-env-function/ ⏳

Comment (nelsonic) [reply lib/aes.ex:87]: Env Encryption Key fetching perf benchmark issue added to backlog: https://github.com/dwyl/fields/issues/85

**Review by nelsonic — APPROVED** [review]
@kgautreaux thank you so much for your efforts on this update. 🥇 
@th0mas thanks for helping get the PR over the line. ⭐

**Comment by nelsonic** [comment]
attempting to run `mix hex.publish` ...

getting the following error:

```
Building docs...
Generating docs...
** (FunctionClauseError) no function clause matching in ExDoc.Retriever.doc_ast/3

    The following arguments were given to ExDoc.Retriever.doc_ast/3:

        # 1
        "text/markdown"

        # 2
        %{}

        # 3
        [file: "lib/helpers.ex", line: 2]

    Attempted function clauses (showing 4 out of 4):

        defp doc_ast(_, :none, _options)
        defp doc_ast(_, :hidden, _options)
        defp doc_ast("text/markdown", %{"en" => doc}, options)
        defp doc_ast(other, %{"en" => _}, _)

    (ex_doc 0.22.5) lib/ex_doc/retriever.ex:167: ExDoc.Retriever.doc_ast/3
    (ex_doc 0.22.5) lib/ex_doc/retriever.ex:214: ExDoc.Retriever.get_module_docs/2
    (ex_doc 0.22.5) lib/ex_doc/retriever.ex:133: ExDoc.Retriever.do_generate_node/3
    (ex_doc 0.22.5) lib/ex_doc/retriever.ex:124: ExDoc.Retriever.generate_node/3
    (elixir 1.12.2) lib/enum.ex:3894: Enum.flat_map_list/2
    (elixir 1.12.2) lib/enum.ex:3895: Enum.flat_map_list/2
    (ex_doc 0.22.5) lib/ex_doc/retriever.ex:43: ExDoc.Retriever.docs_from_modules/2
    (ex_doc 0.22.5) lib/ex_doc.ex:26: ExDoc.generate_docs/3
```

Investigating ... 🔍

**Comment by nelsonic** [comment]
For reference: `elixir -v`:
```
Erlang/OTP 24 [erts-12.0.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]

Elixir 1.12.2 (compiled with Erlang/OTP 24)
```

Tried:
```
mix deps.clean --all
mix deps.get
```

From reading https://github.com/elixir-lang/ex_doc/blob/master/lib/ex_doc.ex#L2
Looks like we just need to add line to `lib/helpers.ex` to inform ExDoc that the file doesn't have `moduledoc` ...

**Comment by nelsonic** [comment]
Looks like the `@moduledoc false` line addition worked. ✅ 
Now I just need to reset my hex.pm local password ... https://medium.com/@wiserfirst/reset-forgotten-hex-local-password-67d88c0c5dc4 🙄

**Comment by nelsonic** [comment]
@kgautreaux `fields@2.8.0` contains your code: https://hex.pm/packages/fields/2.8.0 thanks again! 🎉

**Pull request merged**